### PR TITLE
MB-72716: Fix miscalculation of rollback points

### PR DESCRIPTION
**Pull request by CascadingRadium**
- We currently mark any bolt snapshot that is beyond the expiry window, but within the retention window as `live` and pass it on to the time series mechanism that will protect/retain the number of snapshots configured in scorch - `NumSnapshotsToKeep`
- However when computing the extended boundary, we skip over the first snapshot that was after the expiry window. We basically ignore it and continue to track more snapshots up to the cutoff time determined by the first expired snapshot's boundary checkpoint.Fixed this issue and added a unit test. 

```text
B                      L E                     T
+----------------------+-+---------------------+
|    Retention Window  + |    Expiry Window    |
+----------------------+-+---------------------+ ---> time


- N  = Number Of Snapshots to keep
- D  = Rollback Sampling Interval 
- R  = Rollback Retention Factor
- T  = Current timestamp
- C  = Current set of checkpoints
- E  = Expiry timestamp    = T - [(N - 1) * D]
- B  = Boundary Checkpoint = f(C, R, Lt)
- L  = First expired snapshot
- Lt = L's timestamp
- Nr = Number of snapshots in the retention window
- Ne = Number of snapshots in the expiry window

If (Nr >= N && Ne == 0 && Lt < E):
Then 
    persisted snapshots = [{Nr} - L}]
    protected snapshots ∩ L = ∅
```
- `rollbackRetentionFactor` config validation was broken and errored on valid input instead of invalid. Corrected the condition and added a `0 ≤ r ≤ 1` bounds check.
- Retention settings weren't applied when loading existing snapshots on open, because we opened boltDB before rollback settings were parsed from config. This resulted in the first set of checkpoints to have a timestamp of 0, due to which the rollback points used before the restart were lost. Fixed this issue and added a unit test as well.
- Removed a redundant `closeCh` select after handing a persist off to the introducer, the introducer always completes and closes applies the persist unconditionally, so that branch was dead code.
- Ensured that the persister/merge planner options were parsed exactly once, instead of being scattered across multiple places.
- Shifted rollback config to object construction time instead of setting it when opening the index to ensure that restart scenarios do not use null/zero values for scorch options when loading back the index. 
- Fixes an edge case where we could protect lesser than configured number of snapshots, resulting in deficient rollback checkpoints being used for the next iteration. Added a unit test for this.
- Rollback tests were timing-dependent and flaky, driving real indexing with `time.Sleep()` to simulate scenarios. Fixed them with deterministic tests that capture the same semantic meaning without using sleeps.

### CODE REVIEW

File: index/scorch/persister.go [hunk 1]
```diff
@@ -833,11 +865,8 @@ func (s *Scorch) persistSnapshotDirect(snapshot *IndexSnapshot) (err error) {
 		case s.persists <- persist:
 		}
 
-		select {
-		case <-s.closeCh:
-			return segment.ErrClosed
-		case <-persist.applied:
-		}
+		// blockingly wait until the persist has been applied
+		<-persist.applied
```
[the remark this answers is not in the document]
Comment (CascadingRadium) [reply index/scorch/persister.go:852]: cannot happen as introducer is bound to process it and close applied.

File: index/scorch/persister.go [hunk 2]
```diff
@@ -1594,32 +1560,74 @@ func (s *Scorch) rootBoltSnapshotMetaData() ([]*snapshotMetaData, error) {
...
+	if extraSnapshots <= 0 {
+		return liveSnapshots, nil
+	}
+
+	// if we need extra snapshots, compute an expiration duration
+	// beyond which we will not consider snapshots to be live
+	expirationDuration := time.Duration(extraSnapshots) * s.rollbackSamplingInterval
+	if expirationDuration <= 0 {
+		return liveSnapshots, nil
+	}
```
[the remark this answers is not in the document]
Comment (CascadingRadium) [reply index/scorch/persister.go]: resolved

File: index/scorch/persister.go [hunk 3]
```diff
@@ -1397,21 +1386,20 @@ func newCheckPoints(snapshots map[uint64]time.Time) []*snapshotMetaData {
...
+	liveSnapshots, err := s.getLiveSnapshots()
 	if err != nil {
 		return 0, err
 	}
 
-	if len(persistedSnapshots) <= s.numSnapshotsToKeep {
+	if len(liveSnapshots) <= s.numSnapshotsToKeep {
 		// we need to keep everything
 		return 0, nil
 	}
```
[the remark this answers is not in the document]
Comment (CascadingRadium) [reply index/scorch/persister.go]: done

**Review by Likith101 — APPROVED** [review]

**Review by maneuvertomars — APPROVED** [review]
